### PR TITLE
[WIP] Xcode 14: fix tests

### PR DESCRIPTION
**Pull request by descorp**
# WIP PR

## Changes

* fix SwiftUI app: UIViewControllerRepresentable must be `struct`
* fix tests when fetching BrowserInfo takes too long
* automatically manage available networks for ApplePay

### CODE REVIEW

File: AdyenSwiftUI/Present View Controller/ViewControllerPresenter.swift [hunk 1]
```diff
@@ -35,7 +35,7 @@ import SwiftUI
     }
 
     @available(iOS 13.0, *)
-    internal final class FullScreenView: UIViewControllerRepresentable {
+    internal struct FullScreenView: UIViewControllerRepresentable {
```
Comment (mohammedDehairy) [inline AdyenSwiftUI/Present View Controller/ViewControllerPresenter.swift:38]: why a struct?

File: Tests/Components Tests/Apple Pay/ApplePayComponentTests.swift [hunk 2]
```diff
@@ -354,7 +364,16 @@ class ApplePayComponentTest: XCTestCase {
...
+}
+
+extension XCTestCase {
+    
+    func compareCollections<T: Hashable>(_ lhs: Array<T>, _ rhs: Array<T>) -> Bool {
+        if lhs.count != rhs.count { return false }
+
+        let lhsSet = Set<T>(lhs)
+        let rhsSet = Set<T>(rhs)
+        return lhsSet.intersection(rhsSet).count == lhs.count
```
Comment (mohammedDehairy) [inline Tests/Components Tests/Apple Pay/ApplePayComponentTests.swift:380]: maybe override the `==` operator?

File: AdyenSwiftUI/Present View Controller/ViewControllerPresenter.swift (hunk 1 above)
Comment (descorp) [reply AdyenSwiftUI/Present View Controller/ViewControllerPresenter.swift:38]: Dunno, this is question to Apple 🤷‍♂️
It fails otherwise on Xcode 14 (Xcode 13 is fine)

File: Tests/Components Tests/Apple Pay/ApplePayComponentTests.swift (hunk 2 above)
Comment (descorp) [reply Tests/Components Tests/Apple Pay/ApplePayComponentTests.swift:380]: `Array` is already Equatable. Not sure what you mean?

File: Tests/Components Tests/Apple Pay/ApplePayComponentTests.swift [hunk 3]
```diff
@@ -269,13 +269,24 @@ class ApplePayComponentTest: XCTestCase {
...
+    
+    
+    func testNetworks() {
+        if #available(iOS 15.1, *) {
+            let request = PKPaymentRequest()
+            let collection: [PKPaymentNetwork] = [.discover, .dankort]
+            XCTAssertEqual(collection.count, 2)
+            
+            request.supportedNetworks = collection
+            XCTAssertEqual(request.supportedNetworks.count, 1) // 🤷
```
Comment (mohammedDehairy) [inline Tests/Components Tests/Apple Pay/ApplePayComponentTests.swift]: I don't this comment is helping anybody :-D, except it in making the reader smile  :-D

maybe we should add something to explain

Comment (descorp) [reply Tests/Components Tests/Apple Pay/ApplePayComponentTests.swift]: Sure :)

**Comment by descorp** [comment]
Pending issue with SPM https://github.com/apple/swift-package-manager/issues/5767

File: Tests/Components Tests/Apple Pay/ApplePayComponentTests.swift (hunk 3 above)
Comment (descorp) [reply Tests/Components Tests/Apple Pay/ApplePayComponentTests.swift]: Fixed also [reported here](https://developer.apple.com/forums/thread/715110)

**Review by nauaros — DISMISSED** [review]

File: AdyenSwiftUI/Present View Controller/ViewControllerPresenter.swift [hunk 4]
```diff
@@ -72,8 +72,8 @@ import SwiftUI
                                                 context: UIViewControllerRepresentableContext<FullScreenView>) {
             if context.coordinator.currentlyPresentedViewController != nil {
 
-                dismiss(presenter: presenter, context: context) { [weak self] in
-                    self?.present(viewController: viewController, presenter: presenter, context: context)
+                dismiss(presenter: presenter, context: context) {
```
Comment (nauaros) [inline AdyenSwiftUI/Present View Controller/ViewControllerPresenter.swift:75]: No risk for memory leak here?

File: Tests/Components Tests/Apple Pay/ApplePayComponentTests.swift [hunk 5]
```diff
@@ -269,13 +269,28 @@ class ApplePayComponentTest: XCTestCase {
...
+    
+    func testNetworks() {
+        if #available(iOS 15.1, *) {
+            let request = PKPaymentRequest()
+            let collection: [PKPaymentNetwork] = [.dankort]
+            XCTAssertEqual(collection.count, 1)
+            
+            request.supportedNetworks = collection
+            if #available(iOS 16.0, *) {
+                XCTAssertEqual(request.supportedNetworks.count, 0) // For some reason PKPaymentNetwork ignores dankort
```
Comment (nauaros) [inline Tests/Components Tests/Apple Pay/ApplePayComponentTests.swift:288]: May it be a bug in iOS 16?

File: AdyenSwiftUI/Present View Controller/ViewControllerPresenter.swift (hunk 4 above)
Comment (descorp) [reply AdyenSwiftUI/Present View Controller/ViewControllerPresenter.swift:75]: FullScreenView(self) is a struct, so [it is passed as value type](https://github.com/Wolox/ios-style-guide/blob/master/rules/avoid-struct-closure-self.md).
I am making `present(viewController:presenter:context:)` static, so there should be no is no concerns.

**Review by neelSharma12 — DISMISSED** [review]

File: Tests/Components Tests/Apple Pay/ApplePayComponentTests.swift (hunk 5 above)
Comment (descorp) [reply Tests/Components Tests/Apple Pay/ApplePayComponentTests.swift:288]: I have [reported it to Apple](https://developer.apple.com/forums/thread/715110). Will see..

File: Adyen.xcodeproj/project.pbxproj [hunk 6]
```diff
@@ -5396,6 +5404,7 @@
 		};
 		F9A4A303237323E700D774E8 /* Format and lint */ = {
 			isa = PBXShellScriptBuildPhase;
+			alwaysOutOfDate = 1;
```
Comment (mohammedDehairy) [inline Adyen.xcodeproj/project.pbxproj:5407]: what's this for? :-)

File: Scripts/test-CocoaPods-integration.sh [hunk 7]
```diff
@@ -1,37 +1,51 @@
...
 }
 
-set -e # Any subsequent(*) commands which fail will cause the shell script to exit immediately
+function echo_header {
+  echo " "
+  echo "############# $1 #############"
+}
 
-INCLUDE_WECHAT=true
+INCLUDE_WECHAT=false
```
Comment (mohammedDehairy) [inline Scripts/test-CocoaPods-integration.sh:21]: why is this changed?, we chat should be included by default

File: Scripts/test-CocoaPods-integration.sh [hunk 8]
```diff
@@ -1,37 +1,51 @@
...
+# Any subsequent(*) commands which fail will cause the shell script to exit immediately
+set -eo pipefail
+
 function print_help {
   echo "Test CocoaPods Integration"
   echo " "
   echo "test-CocoaPods-integration [-w]"
   echo " "
   echo "options:"
   echo "-w, --exclude-wechat      exclude wechat module"
```
Comment (mohammedDehairy) [inline Scripts/test-CocoaPods-integration.sh]: this should be `--include-wechat` not `--exclude-wechat`, I think I made a mistake :-)

File: Scripts/test-CocoaPods-integration.sh (hunk 7 above)
Comment (descorp) [reply Scripts/test-CocoaPods-integration.sh:21]: the logic was inverted..

File: Adyen.xcodeproj/project.pbxproj (hunk 6 above)
Comment (descorp) [reply Adyen.xcodeproj/project.pbxproj:5407]: this was causing some odd warning in Xcode 14.

**Comment by nyaray** [comment]
This most likely fixes an issue we started having after a team member updated to Xcode 14 and iOS 16, where we got the first issue in the initial comment of the PR.

**Comment by descorp** [comment]
@nyaray 

I'll address this fix in a separate PR
https://github.com/Adyen/adyen-ios/pull/1007

**Comment by nyaray** [comment]
@descorp appreciate it, I'm just glad that we still haven't launched, that could've been a bad look, hehe.
